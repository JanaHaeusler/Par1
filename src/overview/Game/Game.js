import PropTypes from 'prop-types'
import { useState } from 'react'
import { useHistory } from 'react-router-dom'
import styled from 'styled-components/macro'
import Button from '../../app/Button'
import {
  BinIconDark,
  CancelIconLight,
  DetailsIconDark,
  PencilIconDark,
} from '../../app/Icons/Icons'

Game.propTypes = {
  savedGameProfile: PropTypes.object.isRequired,
  onDelete: PropTypes.func.isRequired,
  onEdit: PropTypes.func.isRequired,
  onDetails: PropTypes.func.isRequired,
}

export default function Game({
  savedGameProfile,
  onDelete,
  onEdit,
  onDetails,
}) {
  const history = useHistory()

  const [isSetToDelete, setIsSetToDelete] = useState(false)
  const { location, date, playersString, winner, shots, _id } = savedGameProfile
  const playerNames = playersString

  return (
    <Card>
      {isSetToDelete || (
        <>
          <SavedGameContent>
            <Location>{location}</Location>
            <Date>{date}</Date>
            <PlayerWrapper>
              <h4>Player(s)</h4>
              <span>{playerNames}</span>
            </PlayerWrapper>
            <WinnerWrapper>
              <h4>Winner(s)</h4>
              <span>{winner}</span>
            </WinnerWrapper>
            <ShotsWrapper>
              <h4>Total Shots</h4>
              <span>{shots}</span>
            </ShotsWrapper>
          </SavedGameContent>
          <ButtonWrapper>
            <ButtonDeleteIcon
              onClick={() => setIsSetToDelete(true)}
              data-testid="button-set-delete"
            >
              <BinIconDark />
            </ButtonDeleteIcon>
            <ButtonEditIcon
              onClick={() => handleEdit(_id)}
              data-testid="button-edit"
            >
              <PencilIconDark />
            </ButtonEditIcon>
            <ButtonDetailsIcon
              onClick={() => handleDetails(_id)}
              data-testid="button-details"
            >
              <DetailsIconDark />
            </ButtonDetailsIcon>
          </ButtonWrapper>
        </>
      )}
      {isSetToDelete && (
        <DeleteField>
          <span>Do you want to delete this game?</span>
          <ButtonDelete
            onClick={() => onDelete(_id)}
            iconComponent={<BinIconDark />}
            text="Delete"
            data-testid="button-delete"
          />
          <ButtonCancel
            main
            onClick={() => setIsSetToDelete(false)}
            iconComponent={<CancelIconLight />}
            text="Cancel"
            data-testid="button-cancel"
          />
        </DeleteField>
      )}
    </Card>
  )

  function handleEdit(id) {
    history.push('/edit')
    onEdit(id)
  }

  function handleDetails(id) {
    history.push('/details')
    onDetails(id)
  }
}

const Card = styled.section`
  margin: 0 20px;
`
const SavedGameContent = styled.div`
  padding: 10px;
  display: grid;
  grid-template-rows: repeat(4, auto);
  grid-template-columns: 1fr 1fr;
  grid-gap: 10px;
  border-radius: 25px 25px 0 0;
  background-color: var(--light);
  font-size: 0.9rem;
`
const Date = styled.div`
  grid-column-start: 2;
  grid-row-start: 1;
  justify-self: end;
`
const Location = styled.div`
  grid-column: 1 / 3;
  grid-row-start: 2;
  text-transform: uppercase;
`
const PlayerWrapper = styled.div`
  grid-column: 1 / 4;
  grid-row-start: 3;
`
const WinnerWrapper = styled.div`
  grid-column-start: 1;
  grid-row-start: 4;
`
const ShotsWrapper = styled.div`
  grid-column-start: 2;
  grid-row-start: 4;
`
const ButtonWrapper = styled.div`
  padding: 5px 0;
  display: flex;
  justify-content: space-evenly;
  align-items: baseline;
  width: 100%;
  border-radius: 0 0 25px 25px;
  background: var(--light-transparent);
`
const ButtonDeleteIcon = styled.button`
  margin: 5px;
  padding: 0;
  display: flex;
  border: none;
  background: none;
`
const ButtonEditIcon = styled.button`
  margin: 5px;
  padding: 0;
  display: flex;
  border: none;
  background: none;
`
const ButtonDetailsIcon = styled.button`
  margin: 5px;
  padding: 0;
  display: flex;
  border: none;
  background: none;
`
const DeleteField = styled.div`
  padding: 10px;
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  gap: 10px 20px;
  place-items: center;
  border-radius: 25px;
  background-color: var(--light);

  span {
    grid-column: 1 / 3;
    text-align: center;
    color: var(--primary);
    font-weight: 800;
  }
`
const ButtonDelete = styled(Button)`
  grid-column-start: 1;
  grid-row-start: 2;
`
const ButtonCancel = styled(Button)`
  grid-column-start: 2;
  grid-row-start: 2;
`
