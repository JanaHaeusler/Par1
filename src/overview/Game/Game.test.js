import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import Game from './Game'

const testProps = {
  savedGameProfile: {
    location: 'City Park',
    date: '2020-12-24',
    winner: 'Sarah',
    shots: '28',
    playersString: 'Sarah, Hannes',
    playersArray: ['Sarah', 'Hannes'],
    scores: {
      Sarah: {
        hole1: '1',
        hole2: '2',
        hole3: '3',
        hole4: '4',
        hole5: '5',
        hole6: '6',
        hole7: '7',
        hole8: '1',
        hole9: '2',
        hole10: '3',
        hole11: '4',
        hole12: '5',
        hole13: '6',
        hole14: '7',
        hole15: '1',
        hole16: '2',
        hole17: '3',
        hole18: '4',
      },
      Hannes: {
        hole1: '7',
        hole2: '6',
        hole3: '5',
        hole4: '4',
        hole5: '3',
        hole6: '2',
        hole7: '1',
        hole8: '7',
        hole9: '6',
        hole10: '5',
        hole11: '4',
        hole12: '3',
        hole13: '2',
        hole14: '1',
        hole15: '7',
        hole16: '6',
        hole17: '5',
        hole18: '4',
      },
    },
    _id: 1,
  },
  onDelete: jest.fn(),
  onEdit: jest.fn(),
  onDetails: jest.fn(),
}

describe('Game', () => {
  it('renders snapshot correctly', () => {
    const { container } = render(<Game {...testProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('shows all game key infos', () => {
    const { getByText } = render(<Game {...testProps} />)
    expect(getByText('City Park')).toBeInTheDocument()
    expect(getByText('2020-12-24')).toBeInTheDocument()
    expect(getByText('Sarah, Hannes')).toBeInTheDocument()
    expect(getByText('Sarah')).toBeInTheDocument()
    expect(getByText('28')).toBeInTheDocument()
  })

  it('has delete button', () => {
    const { getByTestId } = render(<Game {...testProps} />)
    expect(getByTestId('button-set-delete')).toBeInTheDocument()
  })

  it('has edit button', () => {
    const { getByTestId } = render(<Game {...testProps} />)
    expect(getByTestId('button-edit')).toBeInTheDocument()
  })

  it('has details button', () => {
    const { getByTestId } = render(<Game {...testProps} />)
    expect(getByTestId('button-details')).toBeInTheDocument()
  })

  it('renders delete field', () => {
    const { getByText, getByTestId } = render(<Game {...testProps} />)
    user.click(getByTestId('button-set-delete'))
    expect(getByText('Do you want to delete this game?')).toBeInTheDocument()
    expect(getByTestId('button-delete')).toBeInTheDocument()
    expect(getByTestId('button-cancel')).toBeInTheDocument()
  })

  it('calls onDeleteMock by clicking second delete button', () => {
    const onDeleteMock = jest.fn()
    const props = { ...testProps, onDelete: onDeleteMock }
    const { getByTestId } = render(<Game {...props} />)
    user.click(getByTestId('button-set-delete'))
    user.click(getByTestId('button-delete'))
    expect(onDeleteMock).toHaveBeenCalledTimes(1)
  })
})
