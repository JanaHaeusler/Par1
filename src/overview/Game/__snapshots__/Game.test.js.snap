// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders snapshot correctly 1`] = `
.c9 {
  stroke: var(--secondary-dark);
}

.c13 {
  stroke: var(--secondary-dark);
}

.c11 {
  stroke: var(--secondary-dark);
}

.c0 {
  margin: 0 20px;
}

.c1 {
  padding: 10px;
  display: grid;
  grid-template-rows: repeat(4,auto);
  grid-template-columns: 1fr 1fr;
  grid-gap: 10px;
  box-shadow: 0 0 10px var(--box-shadow-green);
  border-radius: 25px 25px 0 0;
  background-color: var(--text-light);
  font-size: 0.9rem;
}

.c3 {
  grid-column-start: 2;
  grid-row-start: 1;
  justify-self: end;
}

.c2 {
  grid-column: 1 / 3;
  grid-row-start: 2;
  text-transform: uppercase;
}

.c4 {
  grid-column: 1 / 4;
  grid-row-start: 3;
}

.c5 {
  grid-column-start: 1;
  grid-row-start: 4;
}

.c6 {
  grid-column-start: 2;
  grid-row-start: 4;
}

.c7 {
  padding: 5px 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  width: 100%;
  border-radius: 0 0 25px 25px;
  background: var(--text-light-transparent);
}

.c8 {
  margin: 5px;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: none;
  background: none;
}

.c10 {
  margin: 5px;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: none;
  background: none;
}

.c12 {
  margin: 5px;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: none;
  background: none;
}

<section
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      City Park
    </div>
    <div
      class="c3"
    >
      2020-12-24
    </div>
    <div
      class="c4"
    >
      <h4>
        Player(s)
      </h4>
      <span>
        Sarah, Hannes
      </span>
    </div>
    <div
      class="c5"
    >
      <h4>
        Winner(s)
      </h4>
      <span>
        Sarah
      </span>
    </div>
    <div
      class="c6"
    >
      <h4>
        Total Shots
      </h4>
      <span>
        28
      </span>
    </div>
  </div>
  <div
    class="c7"
  >
    <button
      class="c8"
      data-testid="button-set-delete"
    >
      <svg
        class="c9"
      >
        bin.svg
      </svg>
    </button>
    <button
      class="c10"
      data-testid="button-edit"
    >
      <svg
        class="c11"
      >
        pencil.svg
      </svg>
    </button>
    <button
      class="c12"
      data-testid="button-details"
    >
      <svg
        class="c13"
      >
        details.svg
      </svg>
    </button>
  </div>
</section>
`;
